Artifacts and how to get them

ShadowCurse opened the pull request:
## Changes
- make `devtool` pull latest artifacts from s3
- add ability to download multiple custom artifacts
- add ability to specify which artifacts to use in tests
- add ability to specify artifacts for A/B tests

## Reason
Artifacts are rebuilt periodically now, so update our tools to be able to pull latest ones.
Additionally, while we are at it, add ability to download and switch between multiple
artifacts. This also opened a path for A/B tests of artifacts as well.

## License Acceptance

By submitting this pull request, I confirm that my contribution is made under
the terms of the Apache 2.0 license. For more information on following Developer
Certificate of Origin and signing off your commits, please check
[`CONTRIBUTING.md`][3].

## PR Checklist

- [ ] I have read and understand [CONTRIBUTING.md][3].
- [ ] I have run `tools/devtool checkbuild --all` to verify that the PR passes
  build checks on all supported architectures.
- [ ] I have run `tools/devtool checkstyle` to verify that the PR passes the
  automated style checks.
- [ ] I have described what is done in these changes, why they are needed, and
  how they are solving the problem in a clear and encompassing way.
- [ ] I have updated any relevant documentation (both in code and in the docs)
  in the PR.
- [ ] I have mentioned all user-facing changes in `CHANGELOG.md`.
- [ ] If a specific issue led to this PR, this PR closes the issue.
- [ ] When making API changes, I have followed the
  [Runbook for Firecracker API changes][2].
- [ ] I have tested all new and changed functionalities in unit tests and/or
  integration tests.
- [ ] I have linked an issue to every new `TODO`.

______________________________________________________________________

- [ ] This functionality cannot be added in [`rust-vmm`][1].

[1]: https://github.com/rust-vmm
[2]: https://github.com/firecracker-microvm/firecracker/blob/main/docs/api-change-runbook.md
[3]: https://github.com/firecracker-microvm/firecracker/blob/main/CONTRIBUTING.md

ShadowCurse committed:
feat(devtool): pull newest test artifacts

Instead of using fixed artifact path for a current FC version
pull newest artifacts from S3. This way we can update them
independently while keeping old versions around.

Unfortunately there is no simple way to query newest
added directory in S3, so do the next best thing of looking
through all files in all directories and determining the `newest`
directory be the `LastModified` status of files inside.

Signed-off-by: Egor Lazarchuk <yegorlz@amazon.co.uk>

ShadowCurse committed:
chore: remove redundant snapshot version check for old FC versions

The snapshot version check was iterating over last several FC versions
pulled from S3, but it was unnecessary since there were no cross FC
checks. Remove checks of old FC versions and only test the current
version. Also remove fixup state from `setup-ci-artifacts.sh` since
it is redundant now.

Signed-off-by: Egor Lazarchuk <yegorlz@amazon.co.uk>

ShadowCurse committed:
feat: update the way test artifacts are downloaded and used

The old setup used `build/img` as the base directory
for artifacts pulled from S3. In S3 the artifacts directory
contains 1 subdirectory per architecture (x86_64 and aarch64).
This was causing final local artifacts path to be `build/img/x86_64`
or `build/img/aarch64`.

This commit changes the structure a to have a separate `build/artifacts`
directory with subdirectories containing different versions of
artifacts. The path to currently used artifacts will be placed in
`build/current_artifacts` file. This make is easy to switch between
multiple versions of artifacts without a need to delete/download them.
This also opens a door for A/B testing of artifacts.

Signed-off-by: Egor Lazarchuk <yegorlz@amazon.co.uk>

ShadowCurse committed:
refactor: specify explicit arguments for ab_test.py

Add arguments names for binary paths for ab_test.py

Signed-off-by: Egor Lazarchuk <yegorlz@amazon.co.uk>

ShadowCurse committed:
feat: ability to download multiple artifacts

Modify `download_ci_artifacts` to accept artifacts paths
as input and download by passing these paths to
`ensure_ci_artifacts`. `ensure_ci_artifacts` will default
to latest s3 artifacts if no args were provided.
`--force` still works and only deletes specified artifacts.

Signed-off-by: Egor Lazarchuk <yegorlz@amazon.co.uk>

ShadowCurse committed:
feat(A/B): add customizable artifacts

Update `ab_test.py` with ability to accept custom artifacts for A and B
runs.

Additionally update `pipeline_perf.py` as well. Now REVISION_A_ARTIFACTS
and REVISION_B_ARTIFACTS environment variables specify custom artifacts
which will be used in A/B test

Signed-off-by: Egor Lazarchuk <yegorlz@amazon.co.uk>

Manciukic commented on tools/devtool:
```diff
@@ -138,6 +142,13 @@ DEFAULT_ARTIFACTS_S3_BUCKET=s3://spec.ccfc.min/firecracker-ci
 # Query default S3 bucket with artifacts and return the most recient path
 get_newest_s3_artifacts() {
   echo $(aws s3 ls $DEFAULT_ARTIFACTS_S3_BUCKET/ --no-sign-request | grep "PRE" | sort -k2 | tail -1 | awk -v bucket="$DEFAULT_ARTIFACTS_S3_BUCKET" '{print bucket "/" $2}')
+  # echo "s3://spec.ccfc.min/firecracker-ci/v1.15"
```
leftover?

Manciukic commented on tools/devtool:
```diff
@@ -737,6 +763,9 @@ cmd_test() {
                 do_kvm_check=0
                 ;;
+            "--no-ci-artifacts-check")
+                do_ci_artifacts_check=0
```
maybe `--no-download-artifacts`?

Also, please add it to the docs for the command

ShadowCurse replied:
added a separate commit with this

Manciukic commented on .buildkite/pipeline_perf.py:
```diff
@@ -132,17 +137,28 @@
+        command.append(pipeline.devtool_download_artifacts(artifacts))
+    # Hack because devtool_test already returns an array, just this array always
+    # has just 1 element
+    command.append(pipeline.devtool_test(devtool_opts, test_script_opts)[0])
```
nit: you could use `extend`

Manciukic commented on tools/devtool:
```diff
@@ -605,6 +605,15 @@ cmd_download_ci_artifacts() {
     done
 }
 
+cmd_set_current_artifacts() {
```
why do we need a separate command when the download one would just do the same when they're already downloaded?

ShadowCurse replied:
Change the logic a bit. Now downloading is separate from setting current artifacts.

Manciukic commented on tools/devtool:
I think we need a way to specify which ones we want to be set as the current, or to be explicit that the last one will be set as current

ShadowCurse replied:
The existing behavior is that we download and use a single version of artifacts. This behavior is preserved with `ensure_current_artifacts_are_set_up`.

ShadowCurse committed:
refactor: split downloading and setting current artifacts

Downloading artifacts is a separate step that should
not mess with currently selected artifacts.

Signed-off-by: Egor Lazarchuk <yegorlz@amazon.co.uk>

ShadowCurse commented on tests/conftest.py:
```diff
@@ -399,7 +398,7 @@ def microvm_factory(request, record_property, results_dir, netns_factory):
                 utils.run_cmd(["dmesg", "-dPx"]).stdout
             )
-            shutil.copy(f"/firecracker/build/img/{platform.machine()}/id_rsa", uvm_data)
+            shutil.copy("/srv/current_artifacts/id_rsa", uvm_data)
```
you are correct, switch to using `ARTIFACT_DIR`

ShadowCurse commented on tools/devtool:
```diff
@@ -426,9 +425,10 @@ cmd_help() {
     echo "        Downloads the CI artifacts used for testing from our S3 bucket. If --force is passed, purges any existing"
-    echo "        artifacts first. Useful for refreshing local artifacts after an update, or if something got messed up."
+    echo "        artifacts first. If specific artifacts are listed, only those will be downloaded; otherwise all artifacts"
+    echo "        are downloaded. Useful for refreshing local artifacts after an update, or if something got messed up."
```
mentioned that is should be S3 path

Manciukic replied:
a path or a URI?

ShadowCurse committed:
chore(test.sh): skip aritfact setup if no current is specified

Some tests like style checks do not require artifacts to be present.

Signed-off-by: Egor Lazarchuk <yegorlz@amazon.co.uk>

ShadowCurse committed:
chore(devtool): update cmd_test help

Add missing descriptions for additional arguments

Signed-off-by: Egor Lazarchuk <yegorlz@amazon.co.uk>

Manciukic commented on tools/devtool:
```diff
@@ -116,7 +116,11 @@ CTR_TEST_RESULTS_DIR="${CTR_FC_ROOT_DIR}/test_results"
+
+# File with a single line specifing the name of the
+# currently used artifacts
+LOCAL_ARTIFACTS_CURRENT_DIR_FILE="build/current_artifacts"
```
> I had this idea initially, but as it turns out symlinks will not work because we will create symlinks outside docker, but tests are run inside, so links will be broken

relative symbolic links will always work: ie `cd build/artifacts; ln -s 20251217_abcde current`.

> no particular reason, I just happened to choose current_artifacts, can change if you prefer the other version

I think it makes it clearer

ShadowCurse committed:
feat: make ensure_current_artifacts_are_set_up a devtool command

The command is used internally in the `devtool` to make sure
there are at lest some current artifacts set up.
This behaviour is also required in the popular docker test,
so convert it to a callable command.

Signed-off-by: Egor Lazarchuk <yegorlz@amazon.co.uk>

ShadowCurse committed:
feat(devtool): specify wanted artifacts for `test` command

Add `--artifacts` option to `devtool test` to simplify changing
between different sets of artifacts.

Signed-off-by: Egor Lazarchuk <yegorlz@amazon.co.uk>

ShadowCurse committed:
chore(devtool): minor adjustments

- rename ensure_current_artifacts_are_set_up to
ensure_current_artifacts to shorten a name a bit
- rename LOCAL_ARTIFACTS_CURRENT_DIR_FILE to
  LOCAL_CURRENT_ARTIFACTS_FILE to make name more descriptive
- Simplify logs regarding artifacts status
- remove ` symbols from cmd_help as bash was interpreting them as actual
  commands to execute
- other minor renamings

Signed-off-by: Egor Lazarchuk <yegorlz@amazon.co.uk>

ShadowCurse commented on tools/devtool:
```diff
@@ -559,28 +612,69 @@ cmd_distclean() {
 
-ensure_ci_artifacts() {
-    if ! command -v aws >/dev/null; then
-      die "AWS CLI not installed, which is required for downloading artifacts for integration tests."
```
I think `ok_or_die "Failed to download artifacts using awscli!"` line serves same function here.

Manciukic commented on tools/test.sh:
```diff
@@ -31,8 +31,14 @@ if [ -f $CGROUP/cgroup.controllers -a -e $CGROUP/cgroup.type ]; then
+  cp -ruvfL $(cat build/current_artifacts) /srv/current_artifacts
+else
+  # The directory must exist for pytest to function
+  mkdir -p /srv/current_artifacts
```
why are we creating it as a dir if it's otherwise a file ?

Manciukic commented on tools/devtool:
```diff
@@ -120,7 +120,7 @@ LOCAL_ARTIFACTS_DIR="build/artifacts"
 # File with a single line specifing the name of the
 # currently used artifacts
-LOCAL_ARTIFACTS_CURRENT_DIR_FILE="build/current_artifacts"
+LOCAL_CURRENT_ARTIFACTS_LINK_PATH="build/current_artifacts"
```
nit: I wouldn't call it a link if it's not one

Manciukic reviewed (commented):
Just a few more bits for small fixes and clarity.

ShadowCurse committed:
chore(test): rename /srv/current_artifacts to /srv/test_artifacts

`current_artifacts` is a name of the FILE that contains a path to
artifacts that should be used for test runs. Same name was used
for a DIRECTORY within devctr that contained artifacts copied
from a build/artifacts... To remove this confusion rename the
directory inside devctr to test_artifacts.

Signed-off-by: Egor Lazarchuk <yegorlz@amazon.co.uk>

ShadowCurse commented on .buildkite/pipeline_docker_popular.py:
```diff
@@ -28,7 +28,7 @@
 pipeline.build_group(
     ":whale: Docker Popular Containers",
     [
-        "./tools/devtool download_ci_artifacts",
```
fixed

JackThomson2 commented on tests/README.md:
```diff
@@ -184,23 +184,30 @@ function to [`.buildkite/pipeline_perf.py`](../.buildkite/pipeline_perf.py). To
-branch and the `HEAD` of your current branch, run
+- _dir A_ and _dir B_ are directories containing firecracker and jailer binaries
+  whose performance characteristics you wish to compare
+- _name A_ and _name B_ are names of arifacts A and B run will to use
```
nit typo

Manciukic committed:
Merge branch 'main' into pull_latest_s3_artifacts
